File: app/src/main/java/com/ngengs/android/popularmovies/apps/adapters/MovieListAdapter.java
package com.ngengs.android.popularmovies.apps.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.ngengs.android.popularmovies.apps.R;
import com.ngengs.android.popularmovies.apps.data.MoviesDetail;
import com.ngengs.android.popularmovies.apps.databinding.ItemMovieListBinding;
import com.ngengs.android.popularmovies.apps.utils.ResourceHelpers;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ngengs on 6/15/2017.
 */

@SuppressWarnings("DefaultFileTemplate")
public class MovieListAdapter extends RecyclerView.Adapter<MovieListAdapter.ViewHolder> {
    @SuppressWarnings("unused")
    private static final String TAG = "MovieListAdapter";
    private final Context context;
    private final List<MoviesDetail> data;
    private final MovieListAdapter.ClickListener clickListener;

    public MovieListAdapter(Context context, MovieListAdapter.ClickListener clickListener) {
        this.context = context;
        this.data = new ArrayList<>();
        this.clickListener = clickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemMovieListBinding binding = ItemMovieListBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        String imageUrl = data.get(position).getPosterPath();
        if (imageUrl != null)
            Picasso.with(context).load(imageUrl).noFade().placeholder(ResourceHelpers.getDrawable(context, R.drawable.ic_collections_white)).into(holder.binding.imagePoster);
        holder.binding.imagePoster.setOnClickListener(view -> holder.itemClick());
        holder.binding.rankPosition.setText(context.getResources().getString(R.string.movie_position, (position + 1)));
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void add(List<MoviesDetail> data) {
        int lastSize = getItemCount();
        this.data.addAll(data);
        notifyItemRangeInserted(lastSize, data.size());
    }

    public void add(MoviesDetail item) {
        data.add(item);
        notifyItemInserted(getItemCount());
    }

    public void deleteById(int movieId) {
        int position = 0;
        for (MoviesDetail item : data) {
            if (item.getId() == movieId) break;
            position++;
        }
        notifyItemRemoved(position);
        data.remove(position);
        Log.d(TAG, "deleteById: position: " + position + " itemCount: " + getItemCount());
        if (position < getItemCount()) notifyItemRangeChanged(position, getItemCount() - position);
    }

    public MoviesDetail get(int position) {
        return data.get(position);
    }

    public List<MoviesDetail> get() {
        return data;
    }

    public void clear() {
        int lastSize = getItemCount();
        data.clear();
        notifyItemRangeRemoved(0, lastSize);
    }

    public interface ClickListener {
        void OnClickListener(int position);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ItemMovieListBinding binding;

        ViewHolder(ItemMovieListBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void itemClick() {
            clickListener.OnClickListener(getAdapterPosition());
        }
    }
}

File: app/src/main/java/com/ngengs/android/popularmovies/apps/adapters/MovieListAdapter.kt
package com.ngengs.android.popularmovies.apps.adapters

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.ngengs.android.popularmovies.apps.R
import com.ngengs.android.popularmovies.apps.data.MoviesDetail
import com.ngengs.android.popularmovies.apps.databinding.ItemMovieListBinding
import com.ngengs.android.popularmovies.apps.utils.ResourceHelpers
import com.squareup.picasso.Picasso

/**
 * Created by rizky.kharisma on 09/11/22.
 * @ngengs
 */
class MovieListAdapter @JvmOverloads constructor(
    private val context: Context,
    private val clickListener: ClickListener,
    private val data: MutableList<MoviesDetail> = mutableListOf(),
) : RecyclerView.Adapter<MovieListAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            ItemMovieListBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val imageUrl = data[position].posterPath
        if (imageUrl != null) Picasso.get().load(imageUrl).noFade()
            .placeholder(ResourceHelpers.getDrawable(context, R.drawable.ic_collections_white))
            .into(holder.binding.imagePoster)
        holder.binding.imagePoster.setOnClickListener { clickListener.onClickListener(position) }
        holder.binding.rankPosition.text =
            context.resources.getString(R.string.movie_position, position + 1)
    }

    override fun getItemCount(): Int = data.size

    fun add(data: List<MoviesDetail>) {
        val lastSize = itemCount
        this.data.addAll(data)
        notifyItemRangeInserted(lastSize, data.size)
    }

    fun add(item: MoviesDetail) {
        data.add(item)
        notifyItemInserted(itemCount)
    }

    fun deleteById(movieId: Int) {
        var position = 0
        for (item in data) {
            if (item.id == movieId) break
            position++
        }
        notifyItemRemoved(position)
        data.removeAt(position)
        Log.d(TAG, "deleteById: position: $position itemCount: $itemCount")
        if (position < itemCount) notifyItemRangeChanged(position, itemCount - position)
    }

    fun get(position: Int): MoviesDetail? = data.getOrNull(position)
    fun get(): List<MoviesDetail> = data

    fun clear() {
        val lastSize = itemCount
        data.clear()
        notifyItemRangeRemoved(0, lastSize)
    }

    interface ClickListener {
        fun onClickListener(position: Int)
    }

    class ViewHolder(val binding: ItemMovieListBinding) : RecyclerView.ViewHolder(binding.root)

    companion object {
        private const val TAG = "MovieListAdapter"
    }
}